homeassistant/components/ping/binary_sensor.py

```diff
@@ -104,32 +105,19 @@ def __init__(self, host, count):
         self.available = False
 
         if sys.platform == "win32":
-            self._ping_cmd = [
-                "ping",
-                "-n",
-                str(self._count),
-                "-w",
-                "1000",
-                self._ip_address,
-            ]
+            self._ping_cmd = f"ping -n {self._count} -w 1000 {self._ip_address}"
         else:
-            self._ping_cmd = [
-                "ping",
-                "-n",
-                "-q",
-                "-c",
-                str(self._count),
-                "-W1",
-                self._ip_address,
-            ]
-
-    def ping(self):
+            self._ping_cmd = f"ping -n -q -c {self._count} -W1 {self._ip_address}"
+
+    async def async_ping(self):
         """Send ICMP echo request and return details if success."""
-        pinger = subprocess.Popen(
-            self._ping_cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE
+        pinger = await asyncio.create_subprocess_shell(
```

Comment:
Please use not the shell one and convert it back to an array. There are some security reasons for this.

Reply:
@pvizeli like here? https://github.com/home-assistant/core/blob/8f9467492d46b6126ac6f9392a25fda68897e8ef/homeassistant/components/shell_command/__init__.py#L57

Reply:
If I understand @pvizeli correctly, we should use `asyncio.create_subprocess_exec` instead of `asyncio.create_subprocess_shell`.

Reply:
@shenxn  like here: https://github.com/home-assistant/core/blob/8f9467492d46b6126ac6f9392a25fda68897e8ef/homeassistant/components/shell_command/__init__.py#L69

So my code will be like this:
```
pinger = await asyncio.create_subprocess_exec(
    *self._ping_cmd,
    stdin=None,
    stdout=asyncio.subprocess.PIPE,
    stderr=asyncio.subprocess.PIPE,
)
```

Reply:
ok, fixed. I've also added more logs, I've looked at shell command integration.